File: Algorithm_234_Palindrome Linked List.cpp
#include <iostream>
#include <stack>
using namespace std;
struct ListNode
{
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};
bool isPalindrome(ListNode* head)
{
	if (!head) return true;
	stack<int> sta;
	ListNode *tmp=head;
	int len=0;
	while (tmp)
	{
		len++;
		tmp=tmp->next;
	}
	tmp=head;
	for (int i=0;i<len;i++)
	{
		if (i<len/2) sta.push(tmp->val);
		else if (i==len/2)
		{
			if (len%2==0)
			{
				if (tmp->val==sta.top()) sta.pop();
				else return false;
			}
		}
		else
		{
			if (tmp->val==sta.top()) sta.pop();
			else return false;
		}
		tmp=tmp->next;
	}
	return sta.empty();
}
int main()
{
	ListNode *n1=new ListNode(1);
	ListNode *n2=new ListNode(0);
	ListNode *n3=new ListNode(0);
	ListNode *n4=new ListNode(1);
	n1->next=n2;
	n2->next=n3;
	n3->next=n4;
	cout<<isPalindrome(n1);
	return 0;
}

File: Algorithm_234_Palindrome Linked List.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def isPalindrome(self, head: ListNode) -> bool:
        if not head or not head.next:
            return True
        slow = fast = head
        mid = ListNode(0)
        while fast and fast.next:
            fast = fast.next.next
            nextSlow = slow.next
            slow.next = mid
            mid = slow
            slow = nextSlow
        cur1 = mid # head of the left list
        cur2 = slow if not fast else slow.next # head of the right list
        # because the tail of the left list is a dummy node ListNode(0)
        # we only need to check whether cur2 is None
        while cur2:
            if cur1.val != cur2.val:
                return False
            cur1, cur2 = cur1.next, cur2.next
        return True